rfc/rfc-46/rfc-46.md

```diff
@@ -84,59 +84,90 @@ is known to have poor performance (compared to non-reflection based instantiatio
+     // HoodieSparkRecord precombine and combineAndGetUpdateValue 
+   }
 
-      @Override
-      public Option<HoodieRecord> combineAndGetUpdateValue(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) {
-         // HoodieSparkRecord combineAndGetUpdateValue
-      }
+   @Override
+   HoodieRecordType getRecordType() {
+     return HoodieRecordType.SPARK;
    }
+}
    
-   /**
-    * Flink-specific implementation 
-    */
-   class HoodieFlinkRecordMerge implements HoodieMerge {
-
-      @Override
-      public HoodieRecord preCombine(HoodieRecord older, HoodieRecord newer) {
-        // HoodieFlinkRecord preCombine
-      }
+/**
+ * Flink-specific implementation 
+ */
+class HoodieFlinkRecordMerger implements HoodieRecordMerger {
+
+   @Override
+   Option<HoodieRecord> merge(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) throws IOException {
+      // HoodieFlinkRecord precombine and combineAndGetUpdateValue 
+   }
 
-      @Override
-      public Option<HoodieRecord> combineAndGetUpdateValue(HoodieRecord older, HoodieRecord newer, Schema schema, Properties props) {
-         // HoodieFlinkRecord combineAndGetUpdateValue
-      }
+   @Override
+   HoodieRecordType getRecordType() {
+      return HoodieRecordType.FLINK;
    }
+}
 ```
 Where user can provide their own subclass implementing such interface for the engines of interest.
 
-#### Migration from `HoodieRecordPayload` to `HoodieMerge`
+#### Migration from `HoodieRecordPayload` to `HoodieRecordMerger`
 
 To warrant backward-compatibility (BWC) on the code-level with already created subclasses of `HoodieRecordPayload` currently
-already used in production by Hudi users, we will provide a BWC-bridge in the form of instance of `HoodieMerge`, that will 
+already used in production by Hudi users, we will provide a BWC-bridge in the form of instance of `HoodieRecordMerger` called `HoodieAvroRecordMerger`, that will 
 be using user-defined subclass of `HoodieRecordPayload` to combine the records.
 
 Leveraging such bridge will make provide for seamless BWC migration to the 0.11 release, however will be removing the performance 
 benefit of this refactoring, since it would unavoidably have to perform conversion to intermediate representation (Avro). To realize
 full-suite of benefits of this refactoring, users will have to migrate their merging logic out of `HoodieRecordPayload` subclass and into
-new `HoodieMerge` implementation.
+new `HoodieRecordMerger` implementation.
+
+Precombine is used to merge records from logs or incoming records; CombineAndGetUpdateValue is used to merge record from log file and record from base file.
+these two merge logics are not exactly the same for some RecordPayload, such as OverwriteWithLatestAvroPaload. 
+We add an Enum in HoodieRecord to mark where it comes from(BASE, LOG or WRITE). `HoodieAvroRecordMerger`'s API will look like following:
```

Comment:
I don't think we even need this: current implementation of `OverwriteWithLatestAvroPayload` is actually implemented inconsistently right now:

- When de-duping it will use `orderingVal` to determine the latest record, but
- When merging persisted record w/ incoming it will just assume that incoming somehow is more recent than persisted one (even though it could be the opposite

We should actually unify its semantic to be consistent across `preCombine` and `combineAndGetUpdateValue`

Reply:
For downward compatibility, I think it is necessary to maintain this difference from OverwriteWithLatestAvroPayload. But in the future implement based on HoodieRecordMerger, this semantics can be unified

Reply:
I agree. But we need to check all the recordPayloads, and the ut involved. Maybe we can do a bug fix later. Because this function is evolving and we can change it.

Reply:
This will be another breaking change then. I totally understand and am with you that it's not making our life easier, but i think it'd be better if we take it upfront make all the breaking changes at once and be done with it.

For ex, this new Enum is going to stick with us for a while if we'd introduce it and may produce new implementations using it which we will need to unwind in the future, which is exactly what i think we'd trying to avoid.

Reply:
I think you're right. I've removed the mark(BASE, LOG or WRITE) in HoodieRecord and unified logic of HoodieSparkRecord.